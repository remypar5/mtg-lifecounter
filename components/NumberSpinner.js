import React from 'react';
import { StyleSheet, View, Text, TextInput } from 'react-native';
import SpinnerButton from './SpinnerButton';

export default class NumberSpinner extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            value: props.value
        };

        this.addPoints = this.addPoints.bind(this);
        this.subtractPoints = this.subtractPoints.bind(this);
    }
    
    render() {
        const props = this.props;
        
        return (
            <View style={ styles.container }>
                <SpinnerButton content="-"
                    onPress={ () => this.subtractPoints(1) } onLongPress={ () => this.subtractPoints(10) } />
                <Text style={ styles.value }>{ this.state.value }</Text>
                <SpinnerButton content="+"
                    onPress={ () => this.addPoints(1) } onLongPress={ () => this.addPoints(10) } />
            </View>
        );
    }
    
    setLife(value) {
        // TODO: use actions and reducers
        this.setState({
            value
        });
    }
    
    addPoints(points = 0) {
        this.setLife(this.state.value + points);
    }
    
    subtractPoints(points = 0) {
        this.setLife(this.state.value - points);
    }
    
}

const styles = StyleSheet.create({
    container: {
        elevation: 20,
        width: '100%',
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'center', // Horizontal
        alignItems: 'stretch', // Vertical
    },
    value: {
        width: '20%',
        height: '100%',
        textAlignVertical: 'center',
        alignSelf: 'center',
        textAlign: 'center',
        color: 'white',
        fontSize: 40,
    }
});