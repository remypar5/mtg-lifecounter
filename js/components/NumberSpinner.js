import React from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, View, Text, TextInput } from 'react-native';
import SpinnerButton from './SpinnerButton';

export default class NumberSpinner extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            value: props.value
        };

        this.increase = this.increase.bind(this);
        this.decrease = this.decrease.bind(this);
    }

    render() {
        const props = this.props;

        return (
            <View style={styles.container}>
                <SpinnerButton content="-"
                    onPress={() => this.decrease(props.step)} onHold={() => this.decrease(props.stepLarge)} />
                <Text style={styles.value}>{this.state.value}</Text>
                <SpinnerButton content="+"
                    onPress={() => this.increase(props.step)} onHold={() => this.increase(props.stepLarge)} />
            </View>
        );
    }

    setValue(value) {
        this.setState({
            value
        });

        this.props.onChange(value);
    }

    increase(points = 0) {
        let newValue = this.state.value + points;

        if (this.props.max !== undefined) {
            newValue = Math.min(this.props.max, newValue);
        }

        this.setValue(newValue);
    }

    decrease(points = 0) {
        let newValue = this.state.value - points;

        if (this.props.min !== undefined) {
            newValue = Math.max(this.props.min, newValue);
        }

        this.setValue(newValue);
    }

}

NumberSpinner.propTypes = {
    value: PropTypes.number.isRequired,
    min: PropTypes.number,
    max: PropTypes.number,
    step: PropTypes.number,
    stepLarge: PropTypes.number,
    onChange: PropTypes.func
};

NumberSpinner.defaultProps = {
    step: 1,
    stepLarge: 10,
    onChange: () => { }
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'center', // Horizontal
        alignItems: 'center', // Vertical
    },
    value: {
        width: '30%',
        height: '100%',
        textAlignVertical: 'center',
        textAlign: 'center',
        alignSelf: 'center',
        color: 'white',
        fontSize: 40,
    }
});