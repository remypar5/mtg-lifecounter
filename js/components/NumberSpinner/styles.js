import { StyleSheet } from 'react-native';
import { normalizeFontSize } from '../../utils';
import { FONT_SIZE_HUGE, FONT_SIZE } from '../../utils/constants';

const styles = StyleSheet.create({
    container: {
        width: '100%',
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'flex-start', // Horizontal
        alignItems: 'center', // Vertical
    },
    valueNormal: {
        flex: 2,
        fontSize: normalizeFontSize(FONT_SIZE),
    },
    valueLarge: {
        flex: 3,
        fontSize: normalizeFontSize(FONT_SIZE_HUGE),
    },
});

export default styles;
