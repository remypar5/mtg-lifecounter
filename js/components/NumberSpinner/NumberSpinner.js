import React from 'react';
import PropTypes from 'prop-types';
import { View, Text } from 'react-native';

import styles from './styles';
import SpinnerButton from '../SpinnerButton';

export default class NumberSpinner extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            value: props.value,
        };

        this.update = this.update.bind(this);
    }

    setValue(value) {
        this.setState({
            value,
        });

        this.props.onChange(value);
    }

    update(points) {
        const { value } = this.state;
        const { min, max } = this.props;

        this.setValue(Math.min(max, Math.max(min, value + points)));
    }

    render() {
        const { step, stepLarge } = this.props;
        const { value } = this.state;
        const { container, button, value: valueStyle } = styles;

        return (
            <View style={container}>
                <SpinnerButton
                    style={button}
                    content="-"
                    onPress={() => this.update(-step)}
                    onHold={() => this.update(-stepLarge)}
                />
                <Text style={valueStyle}>{value}</Text>
                <SpinnerButton
                    style={button}
                    content="+"
                    onPress={() => this.update(+step)}
                    onHold={() => this.update(+stepLarge)}
                />
            </View>
        );
    }
}

NumberSpinner.propTypes = {
    value: PropTypes.number.isRequired,
    min: PropTypes.number,
    max: PropTypes.number,
    step: PropTypes.number,
    stepLarge: PropTypes.number,
    onChange: PropTypes.func,
};

NumberSpinner.defaultProps = {
    step: 1,
    stepLarge: 10,
    min: Number.MIN_SAFE_INTEGER,
    max: Number.MAX_SAFE_INTEGER,
    onChange: () => { },
};
