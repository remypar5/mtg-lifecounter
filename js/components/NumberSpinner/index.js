import NumberSpinner from './NumberSpinner';
import styles from './styles';

export default NumberSpinner;
export {
    styles,
};
