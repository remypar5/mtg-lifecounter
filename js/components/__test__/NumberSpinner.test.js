import React from 'react';
import { shallow } from 'enzyme';

import NumberSpinner from '../NumberSpinner';

describe('<NumberSpinner />', () => {

    let component;

    beforeEach(() => {
        component = shallow(
            <NumberSpinner
                value={20} />
        );
    });

    it('renders without crashing', () => {
        expect(component).toMatchSnapshot();
    });

    it('should initialize with the provided value', () => {
        expect(component
            .find('Text')
            .get(0)
            .props
            .children
        )
        .toBe(20);
    });

    describe('changing the value', () => {
        it('should be able to decrease the value by 1 by pressing the down spinner button', () => {
            expect(component.find('Text').at(0).childAt(0).text()).toBe("20");

            component.find('SpinnerButton').at(0).simulate('press');
            expect(component.find('Text').at(0).childAt(0).text()).toBe("19");
        });

        it('should be able to increase the value by 1 by pressing the up spinner button', () => {
            expect(component.find('Text').at(0).childAt(0).text()).toBe("20");

            component.find('SpinnerButton').at(1).simulate('press');
            expect(component.find('Text').at(0).childAt(0).text()).toBe("21");
        });

        it('should be able to decrease the value by 10 by holding the down spinner button', () => {
            expect(component.find('Text').at(0).childAt(0).text()).toBe("20");

            component.find('SpinnerButton').at(0).simulate('hold');
            expect(component.find('Text').at(0).childAt(0).text()).toBe("10");
        });

        it('should be able to increase the value by 10 by holding the up spinner button', () => {
            expect(component.find('Text').at(0).childAt(0).text()).toBe("20");

            component.find('SpinnerButton').at(1).simulate('hold');
            expect(component.find('Text').at(0).childAt(0).text()).toBe("30");
        });

        it('should call the onChange event listener', () => {
            const onChangeSpy = jest.fn();
            const component = shallow(
                <NumberSpinner
                    value={20}
                    onChange={ onChangeSpy } />
            );

            component.find('SpinnerButton').at(1).simulate('press');
            expect(onChangeSpy).toHaveBeenCalledWith(21);
        });

        describe('limitations', () => {
            beforeEach(() => {
                component = shallow(
                    <NumberSpinner
                        value={20}
                        min={19}
                        max={21} />
                );
            });

            it('should not exceed the provided min value', () => {
                component.find('SpinnerButton').at(0).simulate('press');
                expect(component.find('Text').at(0).childAt(0).text()).toBe("19");

                // Do it again
                component.find('SpinnerButton').at(0).simulate('press');
                expect(component.find('Text').at(0).childAt(0).text()).toBe("19");
            });

            it('should not exceed the provided max value', () => {
                component.find('SpinnerButton').at(1).simulate('press');
                expect(component.find('Text').at(0).childAt(0).text()).toBe("21");

                // Do it again
                component.find('SpinnerButton').at(1).simulate('press');
                expect(component.find('Text').at(0).childAt(0).text()).toBe("21");
            });
        });
    });

});
