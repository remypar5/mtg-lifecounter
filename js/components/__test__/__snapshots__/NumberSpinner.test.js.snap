// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberSpinner /> renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NumberSpinner
    onChange={[Function]}
    step={1}
    stepLarge={10}
    value={20}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <SpinnerButton
          content="-"
          onHold={[Function]}
          onPress={[Function]}
        />,
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "center",
              "color": "white",
              "fontSize": 40,
              "height": "100%",
              "textAlign": "center",
              "textAlignVertical": "center",
              "width": "30%",
            }
          }
        >
          20
        </Text>,
        <SpinnerButton
          content="+"
          onHold={[Function]}
          onPress={[Function]}
        />,
      ],
      "style": Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "content": "-",
          "onHold": [Function],
          "onPress": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "accessible": true,
          "allowFontScaling": true,
          "children": 20,
          "disabled": false,
          "ellipsizeMode": "tail",
          "style": Object {
            "alignSelf": "center",
            "color": "white",
            "fontSize": 40,
            "height": "100%",
            "textAlign": "center",
            "textAlignVertical": "center",
            "width": "30%",
          },
        },
        "ref": null,
        "rendered": 20,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "content": "+",
          "onHold": [Function],
          "onPress": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <SpinnerButton
            content="-"
            onHold={[Function]}
            onPress={[Function]}
          />,
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "alignSelf": "center",
                "color": "white",
                "fontSize": 40,
                "height": "100%",
                "textAlign": "center",
                "textAlignVertical": "center",
                "width": "30%",
              }
            }
          >
            20
          </Text>,
          <SpinnerButton
            content="+"
            onHold={[Function]}
            onPress={[Function]}
          />,
        ],
        "style": Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "content": "-",
            "onHold": [Function],
            "onPress": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "accessible": true,
            "allowFontScaling": true,
            "children": 20,
            "disabled": false,
            "ellipsizeMode": "tail",
            "style": Object {
              "alignSelf": "center",
              "color": "white",
              "fontSize": 40,
              "height": "100%",
              "textAlign": "center",
              "textAlignVertical": "center",
              "width": "30%",
            },
          },
          "ref": null,
          "rendered": 20,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "content": "+",
            "onHold": [Function],
            "onPress": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
